README.md
# Suitest

Suitest is a powerful and easy-to-use JavaScript BDD test suite

* Very simple to use
* Minimum functions and maximum capacity
* Client-side and server-side (including Node.js) support
* One of the most lightweightest libraries for unit testing
* Support for working with asynchronous code
* Outline function callbacks!
* Modular system
* Fluent interface support (chaining)


####BDD practices:

Behavior-driven development (or BDD) is an agile software development technique devised by Dan North as a response to the issues encountered teaching test-driven development.

*The practices of BDD include:*

* Establishing the goals of different stakeholders required for a vision to be implemented
* Drawing out features which will achieve those goals using feature injection
* Involving stakeholders in the implementation process through outside–in software development
* Using examples to describe the behavior of the application, or of units of code
* Automating those examples to provide quick feedback and regression testing
* Using 'should' when describing the behavior of software to help clarify responsibility and allow the software's functionality to be questioned
* Using 'ensure' when describing responsibilities of software to differentiate outcomes in the scope of the code in question from side-effects of other elements of code.
* Using mocks to stand-in for collaborating modules of code which have not yet been written


##Synopsis


*Add a test to run*

```javascript
.test ( name, callback, [, context ] );
```
<br />
*A comparison assertion*

```javascript
.exec ( x, [, y, operator ] );
```
<br />
*Register a callback to fix test result*

```javascript
.done ( [ callback, context] );
```
<br />
*Add test description*

```javascript
.describe ( description );
```
<br />
*Register outline function callbacks*

```javascript
.get ( name );
```
<br />
*Throws an exception when test run and stop all next tests*

```javascript
.stop ();
```
<br />
*A boolean assertion*

```javascript
.is ();
```
<br />
*Register a final callback whenever all the tests have finished running*

```javascript
.finish ( [callback, [total, filed, passed, time] );
```

## Installation

#### Downloading

Choose one of the ways to download the Suitest:

*npm*

```
sudo npm install -g suitest
```

*git*

```
git clone git://github.com/monolithed/Suitest.git
```

*zip/gzip*

```
https://github.com/monolithed/Suitest/downloads
```

#### Server-side including

```javascript
var Suitest = require('./suitest.js');
```

#### Client-side including

```html
<script src="./suitest.js" type="text/javascript"></script>
```

#### Using

```javascript
var unit = new Suitest('Module name');
```

## Screenshot

![Suitest](http://habrastorage.org/storage2/36a/e9f/7b7/36ae9f7b7c575ebbde1323071a1d4216.png "Suitest")

## API


### .test ( name, callback, [, context ] );

```javascript
var unit = new Suitest;

unit.test('test name', function(unit) {
	unit.exec(true, 1); // true
	unit.done();
});
```

### .exec ( x, [, y, context ] );

*Using with one parameter:*

```javascript
unit.exec(true); // true
```

*Using with two parameter:*

```javascript
unit.exec(true, 1); // true, because default operation is ==
```

*Using with three parameters:*

```javascript
unit.exec(true, 1, '==='); // false, because true and 1 are not equivalent
```

### .done ( [ callback ] );

*Simple using:*

```javascript
unit.done();
```

*Testing asynchronous code:*

```javascript
unit.test('test name', function(unit) {
	setTimeout(function() {
		unit.exec(true, 1); // true
		unit.done();
	}, 2000);
});
```

### .describe ( description );

```javascript
unit.describe('Test description');
```

### .finish ( [callback, [total, filed, passed, time] );

```javascript
unit.test('test 1', function() {
    this.exec(true).done();
});

unit.test('test 2', function(unit) {
   this.exec(true).done();
});

unit.finish(function(data) {
    console.log('Total:', total, 'Filed: ', filed, 'Passed: ', passed, 'Time: ', time);
});

// Total: 6, Filed: 2, Passed: 4, Time: 1.00
```

### .stop ();

```javascript
unit.stop();
```

### .is ()

```javascript
unit.exec(true, 1);
unit.is(); // true
```

### .get ( name );

*Outline function callback*

```javascript
var set = function() {
	return unit
		.get('test 6')
		.exec(true, 1)
		.done();
};
```

*Simple using*

```javascript
unit.test('test', function(unit) {
	set(); // true
});
```

*Using with asynchronous code*

```javascript
unit.test('test', function(unit) {
	setTimeout(set, 2000); // true
});
```

*Function.prototype.bind / Function.prototype.call* <br />
Also you can use it without \<get\> method (but this not recommended):

```javascript
var set = function() {
	return this
		.exec(true, 1)
		.done();
};

unit.test('test', function(unit) {
	set.call(unit); // true
});
```

### Context

*There're several ways to set the calling context* <br />

*this*

```javascript
unit.test('test', function() {
	this.exec(true).done();
});
```

*parameter*

```javascript
unit.test('test', function(context) {
	context.exec(true).done();
});
```

*.get()*

```javascript
var set = function() {
	return unit.get('test').exec(true).done();
};

unit.test('test', set);
```

*.call/apply/bind()*

```javascript
var set = function() {
	return this.exec(true).done();
};

unit.test('test', function() {
	set.call(this);
});
```


### Chaining (Fluent interface)


```javascript
var unit = new Suitest;

unit
.test('test 1', function(unit) {
	unit.exec(true, 1).done();
})

.test('test 2', function(unit) {
	unit.describe('test description').exec(true, 1).done();
})

.test('test 2', function(unit) {
	if (!unit.exec(true, 1).is())
		unit.stop();
});
```


### Operators

*`==` operator is used by default and for most operations, you can miss it:*

```javascript
unit.exec(true, 1);
```

*You can also specify it explicitly:*

```javascript
unit.exec(true, 1, '==');
```

*Other available operators:*
<br />
`===`, `!==`, `!==`, `!=`, `<`, `>`, `<=`, `>=`, `eq` <br />

*Special `eg` operator for egal comparison.
The internal comparison abstract operation SameValue(x, y), where x and y are ECMAScript language values, produces true or false (ECMAScript 5 9.12).*
<br />


```javascript
unit.exec(NaN, NaN, 'eq'); // true
unit.exec(0,   0, 'eq');   // false
unit.exec(-0, +0, 'eq');   // false
unit.exec('0', 0, 'eq');   // false
```


## License

* Suitest is licensed under the MIT (MIT_LICENSE.txt) license

* Copyright (c) 2012 [Alexander Guinness] (https://github.com/monolithed)

HISTORY.md
# 0.0.9 / 2012-08-19
* Description section now 63 chars
* \__private\__.stop -> this.__log__.stop
* README: License
* README: call/apply/bind fixed example
* README: Installation (Downloading, including, Using)
* New instance system
* The modules are supported via constructor
* <name> - the module name
* <config> now is a static function
* <exec>: !== -> !=
* Global refactoring

# 0.0.8 / 2012-08-18
* `[critical]` Fixed bug with object instantion
* \__private\__.timeout -> \__config\__.timeout
* Added __private.log.info
* Refactoring

# 0.0.7 / 2012-08-17
* Added \__private\__.extend()
* `[critical]` Fixed bug with Expected and Actual params in the <done>
* `[critical]` Fixed tests

# 0.0.6 / 2012-08-16
* Added Suitest.config
* Added <finish>
* More documentation
* <text> replaced with <describe>

# 0.0.5 / 2012-08-15
* Added safe conditions for console object and setTimeout
* `[critical]` Fixed bug with global object in strict mode (Node.js worked fine o_O)
* `[critical]` Fixed unicode escapes bug for host engines

# 0.0.4 / 2012-08-14
* More documentation
* Refactoring

# 0.0.3 / 2012-08-13

* setTimeout fix;
* done ( [ callback, [, context] ] );
* Getting total elapsed time
* Added \__private\__.time

# 0.0.2 / 2012-08-13

* Added timer for <test> (fix for queue calls)
* HISTORY.md:
* README.md: added image
* npm support: package.json
* Server-side support (NodeJS: module.exports)
* README.md: added image
* Colored text in console
* New method .get() for outline function callbacks

# 0.0.1 / 2012-08-12

* Support for working with asynchronous code
* Fluent interface support (chaining)
* .test( name, callback, [, context ] );
* .exec( x, [, y, context ] );
* .done( [ name ] );
* .text( description );
* .stop();
* .is();


test.js
var Suitest = require('./suitest.js');

var unit = new Suitest;

// test 1
unit.test('test 1', function(unit) {
	unit.describe('Test description 1!').exec(true, 1).done(); // true
});

// test 2
unit.test('test 2', function(unit) {
	unit.describe('Test description 2!');

	if(unit.exec(typeof null !== 'object').is()) // true
		unit.stop();

	unit.done();
});

// test 3
unit.test('test 3', function(unit)
{
	setTimeout(function() {
		unit.exec(Math.acos(-1), Math.PI).done(); // true
	}, 2000);
});

// test 4
unit.test('test 4', function(unit)
{
	setTimeout(function() {
		unit.exec([], Array); // false
		unit.done();
	}, 1000);
});

// test 5
unit.test('test 5', function(unit)
{
	setTimeout(function() {
		unit.exec(1.1 | 0, 1); // true
		unit.done();
	}, 4000);
});


var set = function() {
	return unit
		.get('test 6')
		.exec(true, 1)
		.done();
};

// test 6
unit.test('test 6', function(unit) {
	set(); // true
});

suitest.js
// -*- coding: utf-8; indent-tabs-mode: nil; tab-width: 4; c-basic-offset: 4; -*-

/**
 * Suitest is a powerful and easy-to-use JavaScript test suite
 * @author: Alexander Guinness
 * @version: 0.1.1
 * license: MIT
 * @date: ‎Sun Aug 12 03:30:00 2012
 **/

void function(__object__, __define__)
{
	var __global__ = this,
		__own__ = __object__.hasOwnProperty;

	'use strict';

	var __private__ = {
		info: {
			title:   'Suitest',
			author:  'Alexander Guinnes',
			email:   '<monolithed@gmail.com>',
			version: '0.1.1',
			license: 'MIT',
			year:    2012
		},

		header: function() {
			var info = this.info;

			this.write
			(
				'\n', this.color('gray'), this.line, '\n ',

				// Title
				info.title,

				// Version
				' version: ', info.version,

				// Copyright
				'\n Copyright (c): ', info.author, ', ',

				// Author's e-mail
				info.email, ', ',

				// Release year
				info.year, '\n',

				this.line, this.color('reset')
			);
		},

		/**
		 * __private__.define
		 * It is used to add an own properties and set the descriptors:
		 * {
		 *    __config__urable: false,
		 *    enumerable: false,
		 *    writable: false
		 * }
		 * @param {Object} object
		 * @return {void}
		**/
		define: function(object)
		{
			var __set__ = function(name, value) {
				if (__define__)
					__define__(this, name, {
						value: value,
						writable: true
					});
				else
					this[name] = value;
			}

			for (var key in object) {
				if (__own__.call(object, key))
					__set__.call(this, key, object[key]);
			}
		},

		/**
		 * __private__.color
		 * Getting ASCII codes for coloring output text
		 * @param {String} color - color name
		 * @return {String} Unicode unit
		**/
		color: function(color)
		{
			if (__global__.toString() === '[object Window]')
				return '';

			return '\u001b[' + {
				red    : '31m',
				green  : '32m',
				yellow : '33m',
				blue   : '36m',
				gray   : '37m',
				reset  : '0m'
			}[color];
		},

		/**
		 * __private__.extend
		 * @param {Object} x
		 * @param {Object} y
		 * @return {void}
		**/
		extend: function(x, y) {
			for (var key in y)
				__own__.call(y, key) && (x[key] = y[key]);
		},

		/**
		 * __private__.init
		 * Holds the fist initialization status
		**/
		init: 0,

		/**
		 * __private__.is
		 * Determine the internal ECMAScript [[Class]] of an object.
		 * @param {Object} object
		 * @param {String} name
		 * @return {Boolean}
		**/
		is: function(object, name) {
			return __object__.toString.call(object) === '[object ' + name + ']';
		},

		/**
		 * __private__.line
		 * Dashed lines generator
		 * @return {void}
		**/
		line: Array(64).join('-'),

		/**
		 * __private__.time
		 * Getting total elapsed time
		 * @param {Array} array
		 * @return {Number} elapsed time in seconds
		**/
		time: function(array)
		{
			var result = 0,
			i = array.length >>> 0;

			while (i--)
				result += array[i];

			return result + 'ms';
		},

		/**
		 * __private__.write
		 * Output stream function
		 * @return {void}
		**/
		write: function()
		{
			/*
			* Object console is not part of ECMAScript standard,
			* so we'd check the one.
			* But there're some stupid problems in the engines:
			*
			* Object.prototype.toString.call(console):
			* [object Console] // Chrome
			* [object Object]  // NodeJS, Opera, IE
			*/
			var console = __global__.console;

			if (console && typeof console === 'object')
				console.log(Array.prototype.join.call(arguments, ''));
		}
	};

	/** @public */
	var __config__ = {
		indent: '     ', // Set indentation
		describe: 58,    // Max width for description section
		timeout:  25     // Timeout for the <time> callback
	};

	/**
	 * Object.is
	 * The internal comparison abstract operation SameValue(x, y),
	 * where x and y are ECMAScript language values, produces true or false (ECMAScript 5 9.12).
	 * @param {*} - first generic value for egal comparison
	 * @param {*} - second generic value for egal comparison
	 * @return {Boolean}
	 *
	 * @example:
	 *
	 * Object.is(0,-0)     // false
	 * Object.is('0', 0)   // false
	 * Object.is(0, 0)     // true
	 * Object.is(NaN, NaN) // true
	**/
	__private__.define.call(Object, {
		is: function(x, y)
		{
			// 0 === -0, NaN !== NaN, 0 = false, etc.
			if (x === y)
				return x !== 0 || 1 / x === 1 / y;

			// object !== object ([] !== [], {} !== {}, etc.)
			return x !== x && y !== y;
		}
	});

	/** @constructor */
	__global__.Suitest = function(name)
	{
		if (!(this instanceof Suitest))
			return new Suitest(name);

		/** @static */
		__private__.define.call(this, {
			/**
			 * Suitest.__log__
			 * Holds statistics
			**/
			__log__: {
				name:   name, // The module name
				data:   [],   // <exec> operands (x, y)
				info:   [],   // Result info
				time:   [],   // Elapsed time
				stack:  0,    // The Temporary property to get final callback
				total:  0,    // Total number of tests
				status: 0,    // The Temporary property to get a periodic test status
				failed: 0,    // Total number of failed tests
				passed: 0,    // Total number of passed tests
				params: 0,    // The <exec> parameters,
				context: {}   // <get> { test : context },
			}
		});
	};

	__private__.define.call(Suitest, {
		/**
		 * @static
		 * Suitest.config
		 * Register config
		**/
		config: function(object) {
			if (__private__.is(object, 'Object'))
				__private__.extend(__config__, object);
		}
	});

	__private__.define.call(Suitest.prototype, {
		/**
		 * Suitest.test
		 * Add a test to run
		 * @public
		 * @param {*} name - Test name
		 * @param {Function} callback
		 * @param {Object} context
		 * @return {Object} this
		 *
		 * @example:
		 * var unit = new Suitest;
		 *
		 * // test 1
		 * unit.test('test 1', function(unit) {
		 *    unit.exec(true, 1, '=='); // true
		 *    unit.done();
		 * });
		**/
		test: function(name, callback, context)
		{
			if (!name || !__private__.is(callback, 'Function'))
				throw new TypeError('Suitest.test ( name, callback, [, context ] );');

			// The callbacks will be set as properties for <test>
			var data = {
				name: name,
				done: this.done,
				exec: this.exec,
				stop: this.stop,
				get:  this.get,
				is:   this.is,
				describe: this.describe,
				__log__: this.__log__
			};

			// Set context
			this.__log__.context[name] = data;

			var apply = function()
			{
				// Set start time
				data.time = +new Date;
				callback.call(data || context, data);
			},

			timeout = __global__.setTimeout;

			// Apply callback
			if (__private__.is(timeout, 'Function'))
				timeout(apply, __config__.timeout);

			else apply();

			// Display headline
			if (__private__.init++ === 1)
				__private__.header();

			this.__log__.stack++;
			this.__log__.total++;

			return this;
		},

		/**
		 * Suitest.text
		 * Add test description
		 * @public
		 * @param {String} text - Test description
		 * @return {Object} this
		 *
		 * @example:
		 *
		 * unit.test('test 1', function(unit) {
		 *    unit.text('Compare two values');
		 *    unit.exec(true, 1);
		 *    unit.done();
		 * });
		**/
		describe: function(text) {
			this.__log__.text = text;
			return this;
		},

		/**
		 * Suitest.done
		 * Register a callback to fix test result
		 * @public
		 * @param {Function} [ callback ]
		 * @param {Object} [ context ]
		 * @return {Object} this
		 *
		 * @example:
		 * var unit = new Suitest;
		 *
		 * // test 1
		 * unit.test('test 1', function(unit) {
		 *    unit.exec(true, 1); // true
		 *    unit.done();
		 * });
		 *
		 * // test 2
		 * unit.test('test 2', function(unit) {
		 *    setTimeout(function() {
		 *        unit.exec(true, 1); // true
		 *        unit.done();
		 *    }, 2000);
		 * });
		**/
		done: function(callback, context)
		{
			// Elapsed time
			var time = +new Date - this.time,
				__log__ = this.__log__;

			var text = __log__.text || '',
				values = '';

			// Display the <text> section
			if (text)
				text = '\n' + __config__.indent + 'Description: ' + text;

			// Display the extended statistics if the <exec> passed more than two parameters
			if (__log__.params >= 2)
				values = '\n' + __config__.indent + 'Expected: ' + __log__.data[0] +
						 '\n' + __config__.indent + 'Actual:   ' + __log__.data[1];

			var status = __log__.status;

			// Periodic reports (body)
			__log__.info.push
			(
				// Test name
				'\n', __private__.color('blue'), '<', this.name, '>', __private__.color('reset'),

				// Test description
				text.replace(new RegExp('.{' + __config__.describe + '}', 'g'), '$&\n' + __config__.indent),

				// Extended statistics ( Expected | Actual )
				values,

				// Test status color
				'\n', __config__.indent, 'Status:   ', status == 'passed' ? __private__.color('green') :

				// Test status ( passed | failed )
				__private__.color('red'), status.toUpperCase(), __private__.color('reset'),

				// Elapsed time
				'\n', __config__.indent, 'Time:     ', time, 'ms\n'
			);

			// Keep timers
			__log__.time.push(time);

			// Total statistics by the module
			if (--__log__.stack === 0 || __log__.stop)
			{
				var total_time = __private__.time(__log__.time),
					passed = __log__.passed,
					failed = __log__.failed,
					total  = __log__.total;

				__log__.info.push
				(
					'\n', __private__.color('gray'), __private__.line, '\n',

					// Total number of tests
					' Total: ', total, ' tests, ',

					// Total number of passed tests
					passed, ' passed, ',

					// Total number of failed tests
					failed, ' failed, ',

					// Total time elapsed
					'time: ', total_time, '\n',

					// Line
					__private__.line, __private__.color('reset')
				);

				// Apply <finish> callback
				if (__private__.is(__log__.finish, 'Function'))
				{
					__log__.finish({
						total:  total,
						failed: failed,
						passed: passed,
						time:   total_time
					});

					__log__.finish = null;
				}

				// Add the module name
				__log__.info.unshift('\n[' + __log__.name + ']\n');

				// Show test result
				__private__.write.apply(null, __log__.info);
			}

			// Erase the <text> section
			__log__.text = '';

			// Set default values
			__log__.data = [];

			// The <done> callback
			if (__private__.is(callback, 'Function'))
			{
				callback.call(this || context, {
					status: status,
					time:   time
				});
			}

			// Stop all tests
			if (__log__.stop)
				throw new Error('Stopped test execution!');

			return this;
		},

		/**
		 * Suitest.exec
		 * A comparison assertion
		 * @public
		 * @param {*} x - First operand
		 * @param {*} y - Second operand
		 * @param {String} [ operator ] - Default operation is ==
		 * @return {Object} this
		 *
		 * @example:
		 * var unit = new Suitest;
		 *
		 * // test 1
		 * unit.test('test 1', function(unit) {
		 *    unit.exec(true, 1); // true
		 *    unit.done();
		 * });
		 *
		 * // test 2
		 * unit.test('test 2', function(unit) {
		 *    unit.exec(true, 1, '==='); // false
		 *    unit.done();
		 * });
		 *
		 * // test 3
		 * unit.test('test 3', function(unit) {
		 *    var compare = true == 1;
		 *    unit.exec(compare); // true
		 *    unit.done();
		 * });
		**/
		exec: function(x, y, operator)
		{
			var length = arguments.length,
				status = x;

			if (!length)
				throw new TypeError('Suitest.exec ( x, [, y, context ] );');

			// Apply operations to operands x, y
			if (length >= 2)
			{
				status = {
					'==' : x ==  y,
					'===': x === y,
					'!==': x !== y,
					'!=' : x !=  y,
					'<'  : x  <  y,
					'>'  : x  >  y,
					'<=' : x <=  y,
					'>=' : x >=  y,
					'eg' : Object.is(x, y) // Egal comparison
				}[operator || '=='];
			}

			status = status ? 'passed' : 'failed';

			var __log__ = this.__log__;

			// Set <exec> arguments length
			__log__.params = length;

			// Set test status
			__log__.status = status;

			// Counting status
			__log__[status]++;

			// Set <exec> values
			__log__.data = [x, y];

			return this;
		},

		/**
		 * Suitest.finish
		 * Register a final callback whenever all the tests have finished running
		 * @public
		 * @exports Suitest.finish as __private__.finish
		 * @param {Function} callback - Register for object properties:
		 *    total  - The total number of tests
		 *    filed  - The number of failures,
		 *    passed - The number of tests that passed assertions,
		 *    time   - The total time in milliseconds for all tests
		 *
		 * @return {Object} this
		 *
		 * @example:
		 *
		 * unit.test('test 1', function() {
		 *     this.exec(true).done();
		 * });
		 *
		 * unit.test('test 2', function(unit) {
		 *    this.exec(true).done();
		 * });
		 *
		 * unit.finish(function(data) {
		 *     console.log('Total:', total, 'Filed: ', filed, 'Passed: ', passed, 'Time: ', time);
		 * });
		 *
		 * // Total: 6, Filed: 2, Passed: 4, Time: 1.00
		 *
		**/
		finish: function(callback) {
			this.__log__.finish = callback;
			return this;
		},

		/**
		 * Suitest.get
		 * Register outline function callbacks
		 * @public
		 * @param {String} name
		 * @return {Object} { test : context }
		 *
		 * @example:
		 *
		 * // Simple using
		 * var set = function() {
		 *     return unit
		 *        .get('test')
		 *        .exec(true, 1)
		 *        .done();
		 * };
		 *
		 * unit.test('test', function(unit) {
		 *    set(); // true
		 * });
		 *
		 * // Using with asynchronous code
		 * var set = function() {
		 *     return unit
		 *        .get('test')
		 *        .exec(true, 1)
		 *        .done();
		 * };
		 *
		 * unit.test('test', function(unit) {
		 *    setTimeout(set, 2000); // true
		 * });
		**/
		get: function(name) {
			return this.__log__.context[name];
		},

		/**
		 * Suitest.is
		 * A boolean assertion
		 * @public
		 * @return {Boolean}
		 *
		 *  @example:
		 *
		 * // Simple using
		 * unit.test('test 1', function(unit) {
		 *    unit.exec(true, 1);
		 *    if (!unit.is())
		 *      unit.stop();
		 * });
		 *
		 * // Convenient using
		 * unit.test('test 2', function(unit) {
		 *    if (!unit.exec(true, 1).is())
		 *      unit.stop();
		 * });
		**/
		is: function() {
			return this.__log__.status == 'passed';
		},

		/**
		 * Suitest.stop
		 * Throws an exception when test run and stop all next tests
		 * @public
		 * @return {Object} this
		 *
		 *  @example:
		 *
		 * unit.test('test 1', function(unit) {
		 *    if (x > y)
		 *      unit.stop();
		 * });
		**/
		stop: function() {
			this.__log__.stop = true;
			return this;
		}
	});

	return Suitest;

}(Object.prototype, Object.defineProperty);

// server-side support
try { module.exports = Suitest; } catch(error) {}